---
title: "Genomic Heatmaps and Profiles"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/Genomic_HeatmapsAndProfiles/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style



---


```{r setup_ggplot2, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


---
## Heatmaps and Profiles


```{r download_Data,cache=TRUE,warning=FALSE,message=FALSE}
require(GEOquery)

GSM4332942_BigWig <- getGEOSuppFiles("GSM4332942",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332937_BigWig <- getGEOSuppFiles("GSM4332937",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332947_BigWig <- getGEOSuppFiles("GSM4332947",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332952_BigWig <- getGEOSuppFiles("GSM4332952",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")


GSM4332935_BigWig <- getGEOSuppFiles("GSM4332935",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332940_BigWig <- getGEOSuppFiles("GSM4332940",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332945_BigWig <- getGEOSuppFiles("GSM4332945",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332950_BigWig <- getGEOSuppFiles("GSM4332950",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")


GSM4332934_BigWig <- getGEOSuppFiles("GSM4332934",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332939_BigWig <- getGEOSuppFiles("GSM4332939",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332944_BigWig <- getGEOSuppFiles("GSM4332944",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

GSM4332949_BigWig <- getGEOSuppFiles("GSM4332949",fetch_files = TRUE,
                                     makeDirectory = FALSE,filter_regex = "*.bw")

```


---
## ee

```{r Review_Table,cache=TRUE,warning=FALSE,message=FALSE}
myTableOfBigWigDownloads <- do.call(rbind,list(GSM4332942_BigWig,GSM4332937_BigWig,
                   GSM4332947_BigWig,GSM4332952_BigWig))

```

---
## Peaks

```{r importSupplmentary,cache=TRUE,warning=FALSE,message=FALSE}
require(rio)
myRes <- import_list("~/Downloads/1-s2.0-S1550413120301236-mmc5.xlsx",which=c("Consensus ZBTB1 Peaks","Consensus ATF4 Peaks"))
require(GenomicRanges)
ATF4peaks <- myRes[[1]][-1,1:3]
ZBTB1peaks <- myRes[[2]][-1,1:3]
colnames(ATF4peaks) <- colnames(ZBTB1peaks) <- c("seqnames","start","end")

ATF4GR <- GRanges(ATF4peaks)
names(ATF4GR) <- as.character(ATF4GR)
ZBTB1GR <- GRanges(ZBTB1peaks)
names(ZBTB1GR) <- as.character(ZBTB1GR)
rtracklayer::export(ATF4GR,con="ATF4Peaks.bed")
rtracklayer::export(ZBTB1GR,con="ZBTB1Peaks.bed")
```


---
## Peaks
```{r soggi_ZBTB1,cache=TRUE,warning=FALSE,message=FALSE}
require(profileplyr)
myProf <- BamBigwig_to_chipProfile(c(rownames(GSM4332935_BigWig),
                                     rownames(GSM4332945_BigWig),
                                     rownames(GSM4332940_BigWig),
                                     rownames(GSM4332950_BigWig)),
                                   testRanges = "ATF4Peaks.bed",
                                   style = "point",
                                   format = "bigwig",distanceAround = 2000)

profileObject <- as_profileplyr(myProf)
# hclust <- clusterRanges(profileObject, 
#                         fun = rowSums, 
#                         cutree_rows = , scaleRows = FALSE,
#                         silent = TRUE)

blacklist <- rtracklayer::import("~/Downloads/ENCFF001TDO.bed.gz")
generateEnrichedHeatmap(profileObject)
generateEnrichedHeatmap(profileObject[!profileObject %over% blacklist])
```

```{r soggi_ZBTB1,cache=TRUE,warning=FALSE,message=FALSE}
require(profileplyr)
myProf <- BamBigwig_to_chipProfile(c(rownames(GSM4332935_BigWig),
                                     rownames(GSM4332940_BigWig),
                                     rownames(GSM4332934_BigWig),
                                     rownames(GSM4332939_BigWig)),
                                   testRanges = "ATF4Peaks.bed",
                                   style = "point",
                                   format = "bigwig",distanceAround = 10000)


profileObject <- as_profileplyr(myProf)
hclust <- clusterRanges(profileObject,
                        fun = rowSums,
                        cutree_rows = 2, scaleRows = TRUE,
                        silent = FALSE)

grouped_hclust <- profileplyr::groupBy(hclust,group = ATF4ANDZBTB1peaks_GR,include_nonoverlapping = TRUE)
blacklist <- rtracklayer::import("~/Downloads/ENCFF001TDO.bed.gz")
generateEnrichedHeatmap(hclust[!hclust %over% blacklist])
generateEnrichedHeatmap(grouped_hclust[!grouped_hclust %over% blacklist])
generateEnrichedHeatmap(profileObject[!profileObject %over% blacklist])
require(ggplot2)
grouped_hclust %>% 
  summarize(fun=rowSums,output = "long") %>% 
  ggplot(aes(y=log2(Signal),x=Sample,fill=GR_overlap_names))+geom_boxplot()+coord_flip()

generateEnrichedHeatmap(profileObject[!profileObject %over% ATF4ANDZBTB1peaks_GR])
```

```{r soggi,cache=TRUE,warning=FALSE,message=FALSE}

myRes <- import_list("~/Downloads/1-s2.0-S1550413120301236-mmc4.xlsx")
require(GenomicRanges)
ATF4ANDZBTB1peaks <- myRes[[1]][-1,1:3]
# ZBTB1peaks <- myRes[[2]][-1,1:3]
colnames(ATF4ANDZBTB1peaks) <- colnames(ZBTB1peaks) <- c("seqnames","start","end")

ATF4ANDZBTB1peaks_GR <- GRanges(ATF4ANDZBTB1peaks)
names(ATF4ANDZBTB1peaks_GR) <- as.character(ATF4ANDZBTB1peaks)
rtracklayer::export(ATF4ANDZBTB1peaks_GR,con="ATF4ANDZBTB1peaks_GR.bed")

```


```{r}
c(ENCFF234LPQ="H3K36me3_Heart",
  ENCFF004EZJ="H3K36me3_Liver",
  ,
  ,
  ,
  )
dir("/Users/thomascarroll/Downloads/forPPcourse",full.names=TRUE)
require(profileplyr)
require(magrittr)
require(TxDb.Mmusculus.UCSC.mm10.knownGene)
require(rtracklayer)
genePos <- TxDb.Mmusculus.UCSC.mm10.knownGene %>% genes
export.bed(genePos,"genePos.bed")
myProf <- dir("/Users/thomascarroll/Downloads/forPPcourse",full.names=TRUE)[1] %>%  
                                   BamBigwig_to_chipProfile(
                                   testRanges = "genePos.bed",
                                   style = "percentOfRegion",
                                   format = "bigwig",distanceAround = 100)
myProf2 <- dir("/Users/thomascarroll/Downloads/forPPcourse",full.names=TRUE)[4] %>%  
                                   BamBigwig_to_chipProfile(
                                   testRanges = "genePos.bed",
                                   style = "percentOfRegion",
                                   format = "bigwig",distanceAround = 100)

myProf3 <- dir("/Users/thomascarroll/Downloads/forPPcourse",full.names=TRUE)[2] %>%  
                                   BamBigwig_to_chipProfile(
                                   testRanges = "genePos.bed",
                                   style = "percentOfRegion",
                                   format = "bigwig",distanceAround = 100)

ENCFF827YAZ
profileObject <- as_profileplyr(c(myProf2,myProf3))
hclust <- clusterRanges(profileObject,
                        fun = rowSums,
                        kmeans_k = 4, scaleRows = FALSE,
                        silent = FALSE,clustering_distance_rows = "correlation")

mm10BL <- import.bed("~/Downloads/mm10.blacklist.bed")
generateEnrichedHeatmap(profileObject[!profileObject %over% mm10BL])
generateEnrichedHeatmap(hclust[!hclust %over% mm10BL])

```
# Contact

Any suggestions, comments, edits or questions (about content or the slides themselves) please reach out to our [GitHub](https://github.com/RockefellerUniversity/Genomic_HeatmapsAndProfiles/issues) and raise an issue.
